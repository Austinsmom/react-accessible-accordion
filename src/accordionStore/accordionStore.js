// @flow

import { observable, action } from 'mobx';

export type Item = {
    itemkey: string | number,
    itemUuid: number,
    expanded: boolean,
    disabled: boolean,
};
export type Store = {
    items: Array<Item>,
    accordion: boolean,
    onChange: Function,
    addItem: Item => void,
    removeItem: ($PropertyType<Item, 'itemkey'>) => void,
    setExpanded: ($PropertyType<Item, 'itemkey'>, boolean) => void,
};

// eslint-disable-next-line import/prefer-default-export
export const createAccordionStore = ({
    accordion,
    onChange,
}: {
    accordion: boolean,
    onChange: Function,
}): Store => {
    const store: Store = observable({
        items: [],
        accordion,
        onChange,
        addItem: action.bound(function addItem(item) {
            this.items = [...this.items, item];
        }),
        removeItem: action.bound(function removeItem(key) {
            this.items = this.items.filter(item => item.itemkey !== key);
        }),
        setExpanded: action.bound(function setExpanded(
            key: string | number,
            expanded: boolean,
        ) {
            this.items = this.items.map(item => {
                if (item.itemkey === key) {
                    return {
                        ...item,
                        expanded,
                    };
                }
                return item;
            });
        }),
    });
    return store;
};
