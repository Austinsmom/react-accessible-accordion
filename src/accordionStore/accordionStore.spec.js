// @flow

import { createAccordionStore } from './accordionStore';

describe('accordionStore', () => {
    describe('createAccordionStore', () => {
        it('initializes a new store', () => {
            const store = createAccordionStore({
                accordion: true,
                onChange: jest.fn(),
            });
            expect(store).toMatchSnapshot();
        });

        it('can add a new item', () => {
            const store = createAccordionStore({
                accordion: true,
                onChange: jest.fn(),
            });

            expect(store.items.length).toEqual(0);
            store.addItem({
                uuid: 'foo',
                expanded: true,
                disabled: false,
            });
            expect(store.items.length).toEqual(1);
        });

        it('can remove an item after adding some items', () => {
            const store = createAccordionStore({
                accordion: true,
                onChange: jest.fn(),
            });

            expect(store.items.length).toEqual(0);
            store.addItem({
                uuid: 'foo',
                expanded: true,
                disabled: false,
            });

            expect(store.items.length).toEqual(1);
            store.addItem({
                uuid: 'bar',
                expanded: true,
                disabled: false,
            });

            expect(store.items.length).toEqual(2);
            store.removeItem('foo');
            expect(store.items.length).toEqual(1);
        });

        it('can modify the expanded property of an item', () => {
            const store = createAccordionStore({
                accordion: true,
                onChange: jest.fn(),
            });

            store.addItem({
                uuid: 'foo',
                expanded: true,
                disabled: false,
            });

            store.addItem({
                uuid: 'bar',
                expanded: true,
                disabled: false,
            });

            store.setExpanded('foo', false);
            expect(store.items.filter(item => item.expanded).length).toEqual(1);
        });

        it('will collapse any open item if accordion=true', () => {
            const store = createAccordionStore({
                accordion: true,
                onChange: jest.fn(),
            });

            store.addItem({
                uuid: 'foo',
                expanded: true,
                disabled: false,
            });

            store.addItem({
                uuid: 'bar',
                expanded: false,
                disabled: false,
            });

            store.setExpanded('bar', true);
            expect(store.items.filter(item => item.expanded).length).toEqual(1);
        });

        it("won't collapse any open item if accordion=false", () => {
            const store = createAccordionStore({
                accordion: false,
                onChange: jest.fn(),
            });

            store.addItem({
                uuid: 'foo',
                expanded: true,
                disabled: false,
            });

            store.addItem({
                uuid: 'bar',
                expanded: false,
                disabled: false,
            });

            store.setExpanded('bar', true);
            expect(store.items.filter(item => item.expanded).length).toEqual(2);
        });

        it('closes expanded items when expanded item is added and accordion=true', () => {
            const store = createAccordionStore({
                accordion: true,
                onChange: jest.fn(),
            });

            store.addItem({
                uuid: 'foo',
                expanded: true,
                disabled: false,
            });

            store.addItem({
                uuid: 'bar',
                expanded: true,
                disabled: false,
            });

            expect(store.items.filter(item => item.expanded).length).toEqual(1);
        });
    });
});
