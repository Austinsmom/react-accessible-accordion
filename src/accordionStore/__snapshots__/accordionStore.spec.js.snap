// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`accordionStore createAccordionStore initializes a new store 1`] = `
Object {
  "accordion": true,
  "items": Array [],
  "onChange": [MockFunction],
}
`;
